File: nodejs/src/executionEngine.ts
import { randomBytes } from "node:crypto";
import { EventType } from "./events.js";
import {
  SqubeApprovalPendingError,
  SqubeBlockedError,
  SqubeDeniedError,
  SqubeGuardError,
} from "./errors.js";
import { hashPayload } from "./hashing.js";
import { Ledger } from "./ledger.js";
import { defaultPolicy } from "./policy.js";
import { redactValue } from "./redaction.js";
import { promptCliApproval } from "./approval.js";
import {
  ActionStatus,
  Decision,
  type ApprovalFn,
  type OnErrorMode,
  type PolicyFn,
} from "./types.js";

export type ApprovalMode = "sync" | "deferred";

export interface ExecutionContext {
  executionId: string;
  agentId: string;
  action: string;
  resource: string | null;
  parameters: Record<string, unknown>;
  timestamp?: string;
  correlationId?: string;
  rootExecutionId?: string;
}

function utcNowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function newApprovalId(): string {
  return `sq_apr_${randomBytes(6).toString("hex")}`;
}

export class ExecutionEngine {
  readonly ledger: Ledger;
  private policy: PolicyFn;
  private approvalMode: ApprovalMode;
  private approvalTimeoutSeconds: number;
  private approvalFn: ApprovalFn;
  private onError: OnErrorMode;

  constructor(opts: {
    policy?: PolicyFn;
    ledgerPath?: string;
    approvalMode?: ApprovalMode;
    approvalTimeoutSeconds?: number;
    approvalFn?: ApprovalFn;
    onError?: OnErrorMode;
  } = {}) {
    this.ledger = new Ledger(opts.ledgerPath ?? "sqube_ledger.sqlite3");
    this.policy = opts.policy ?? defaultPolicy;
    this.approvalMode = opts.approvalMode ?? "sync";
    this.approvalTimeoutSeconds = opts.approvalTimeoutSeconds ?? 300;
    this.approvalFn = opts.approvalFn ?? promptCliApproval;
    this.onError = opts.onError ?? "fail_closed";
  }

  simulate(ctx: ExecutionContext): Decision {
    return this.policy(ctx.action, ctx.resource, ctx.agentId, {
      parameters: ctx.parameters,
    });
  }

  policyName(): string {
    return this.policy.name || "anonymous_policy";
  }

  async runControlled<T>(ctx: ExecutionContext, fn: () => T | Promise<T>): Promise<T> {
    const ts = ctx.timestamp ?? utcNowIso();
    const rootId = ctx.rootExecutionId ?? ctx.executionId;
    const correlationId = ctx.correlationId ?? rootId;
    const paramsHash = hashPayload(ctx.parameters);
    const summary = redactValue(ctx.parameters);

    this.ledger.insert({
      execution_id: ctx.executionId,
      created_at: ts,
      agent_id: ctx.agentId,
      action: ctx.action,
      resource: ctx.resource,
      parameters_hash: paramsHash,
      parameters_summary: summary,
      decision: Decision.ALLOW,
      policy_id: this.policy.name || "policy",
      status: ActionStatus.REQUESTED,
    });
    this.ledger.appendEvent(
      ctx.executionId,
      EventType.ACTION_REQUESTED,
      ctx.agentId,
      { action: ctx.action, resource: ctx.resource },
      ts
    );

    this.ledger.transitionExecution(
      ctx.executionId,
      ActionStatus.REQUESTED,
      ActionStatus.EVALUATED
    );
    this.ledger.appendEvent(
      ctx.executionId,
      EventType.CONTEXT_RESOLVED,
      ctx.agentId,
      { correlation_id: correlationId, root_execution_id: rootId },
      utcNowIso()
    );

    let decision: Decision;
    try {
      decision = this.simulate(ctx);
    } catch (err) {
      if (this.onError === "fail_open") {
        return this.beginExecute(ctx, fn);
      }
      const message = err instanceof Error ? err.message : String(err);
      this.ledger.transitionExecution(
        ctx.executionId,
        ActionStatus.EVALUATED,
        ActionStatus.FAILED,
        {
          error_message: `policy_error: ${message}`,
          completed_at: utcNowIso(),
        }
      );
      throw new SqubeGuardError(`Policy evaluation failed: ${message}`);
    }
    this.ledger.updateStatus(ctx.executionId, { decision });
    this.ledger.appendEvent(
      ctx.executionId,
      EventType.POLICY_EVALUATED,
      "policy",
      { decision },
      utcNowIso()
    );

    if (decision === Decision.BLOCK) {
      this.ledger.transitionExecution(
        ctx.executionId,
        ActionStatus.EVALUATED,
        ActionStatus.BLOCKED,
        { completed_at: utcNowIso() }
      );
      throw new SqubeBlockedError(ctx.executionId, ctx.action);
    }

    if (decision === Decision.REQUIRE_APPROVAL) {
      if (this.approvalMode === "deferred") {
        const approvalId = newApprovalId();
        const now = utcNowIso();
        const expires = new Date(Date.now() + this.approvalTimeoutSeconds * 1000)
          .toISOString()
          .replace(/\.\d{3}Z$/, "Z");
        this.ledger.transitionExecution(
          ctx.executionId,
          ActionStatus.EVALUATED,
          ActionStatus.WAITING_APPROVAL
        );
        this.ledger.createApprovalRequest(
          approvalId,
          ctx.executionId,
          paramsHash,
          now,
          expires
        );
        this.ledger.appendEvent(
          ctx.executionId,
          EventType.APPROVAL_REQUESTED,
          ctx.agentId,
          { approval_id: approvalId, expires_at: expires },
          now
        );
        throw new SqubeApprovalPendingError(ctx.executionId, approvalId);
      }
      return this.syncApprovalPath(ctx, fn, paramsHash, summary);
    }

    return this.beginExecute(ctx, fn);
  }

  cancelExecution(executionId: string, reason = "cancelled"): void {
    const row = this.ledger.getExecution(executionId);
    if (!row) throw new SqubeGuardError(`unknown execution ${executionId}`);
    const now = utcNowIso();
    if (row.status === ActionStatus.WAITING_APPROVAL) {
      this.ledger.transitionExecution(
        executionId,
        ActionStatus.WAITING_APPROVAL,
        ActionStatus.CANCELLED,
        { completed_at: now, approval_reason: reason }
      );
      return;
    }
    if (row.status === ActionStatus.APPROVED) {
      this.ledger.transitionExecution(
        executionId,
        ActionStatus.APPROVED,
        ActionStatus.CANCELLED,
        { completed_at: now, approval_reason: reason }
      );
      return;
    }
    throw new SqubeGuardError(`cannot cancel execution in status ${row.status}`);
  }

  private async syncApprovalPath<T>(
    ctx: ExecutionContext,
    fn: () => T | Promise<T>,
    paramsHash: string,
    summary: string
  ): Promise<T> {
    this.ledger.transitionExecution(
      ctx.executionId,
      ActionStatus.EVALUATED,
      ActionStatus.WAITING_APPROVAL
    );
    const { approved, approvedBy, reason } = await this.approvalFn({
      executionId: ctx.executionId,
      agentId: ctx.agentId,
      action: ctx.action,
      resource: ctx.resource,
      summary,
      timeoutSeconds: this.approvalTimeoutSeconds,
    });
    const now = utcNowIso();
    if (!approved) {
      const toStatus =
        reason === "timeout" ? ActionStatus.EXPIRED : ActionStatus.DENIED;
      this.ledger.transitionExecution(ctx.executionId, ActionStatus.WAITING_APPROVAL, toStatus, {
        approval_reason: reason,
        completed_at: now,
      });
      throw new SqubeDeniedError(ctx.executionId, reason ?? "denied");
    }
    this.ledger.transitionExecution(ctx.executionId, ActionStatus.WAITING_APPROVAL, ActionStatus.APPROVED, {
      approved_by: approvedBy,
    });
    return this.beginExecute(ctx, fn, ActionStatus.APPROVED);
  }

  async resumeAfterApproval<T>(
    ctx: ExecutionContext,
    fn: () => T | Promise<T>,
    approvalId: string
  ): Promise<T> {
    const row = this.ledger.getExecution(ctx.executionId);
    if (!row) throw new SqubeGuardError(`unknown execution ${ctx.executionId}`);
    if (row.status === ActionStatus.WAITING_APPROVAL) {
      throw new SqubeApprovalPendingError(ctx.executionId, approvalId);
    }
    if (
      row.status === ActionStatus.DENIED ||
      row.status === ActionStatus.EXPIRED ||
      row.status === ActionStatus.BLOCKED
    ) {
      throw new SqubeDeniedError(ctx.executionId, row.status);
    }
    if (hashPayload(ctx.parameters) !== row.parameters_hash) {
      throw new SqubeGuardError("parameters changed after authorization");
    }
    const approval = this.ledger.getApproval(approvalId);
    if (!approval || approval.execution_id !== ctx.executionId) {
      throw new SqubeGuardError(`unknown approval ${approvalId}`);
    }
    const now = utcNowIso();
    if (approval.expires_at && approval.expires_at < now) {
      throw new SqubeDeniedError(ctx.executionId, "approval_expired");
    }
    this.ledger.consumeApproval(approvalId, ctx.executionId, now);
    this.ledger.appendEvent(
      ctx.executionId,
      EventType.ACTION_STARTED,
      ctx.agentId,
      { approval_id: approvalId },
      now
    );
    return this.runFnBody(ctx, fn);
  }

  private async beginExecute<T>(
    ctx: ExecutionContext,
    fn: () => T | Promise<T>,
    fromStatus: ActionStatus = ActionStatus.EVALUATED
  ): Promise<T> {
    this.ledger.transitionExecution(ctx.executionId, fromStatus, ActionStatus.EXECUTING);
    this.ledger.appendEvent(
      ctx.executionId,
      EventType.ACTION_STARTED,
      ctx.agentId,
      {},
      utcNowIso()
    );
    return this.runFnBody(ctx, fn);
  }

  private async runFnBody<T>(ctx: ExecutionContext, fn: () => T | Promise<T>): Promise<T> {
    const start = performance.now();
    try {
      const result = await fn();
      const durationMs = Math.round(performance.now() - start);
      const doneAt = utcNowIso();
      this.ledger.transitionExecution(
        ctx.executionId,
        ActionStatus.EXECUTING,
        ActionStatus.SUCCEEDED,
        {
          result_status: "SUCCESS",
          duration_ms: durationMs,
          completed_at: doneAt,
        }
      );
      this.ledger.appendEvent(
        ctx.executionId,
        EventType.ACTION_SUCCEEDED,
        ctx.agentId,
        {},
        doneAt
      );
      return result;
    } catch (err) {
      const durationMs = Math.round(performance.now() - start);
      const message = err instanceof Error ? err.message : String(err);
      const failedAt = utcNowIso();
      this.ledger.transitionExecution(
        ctx.executionId,
        ActionStatus.EXECUTING,
        ActionStatus.FAILED,
        {
          result_status: "FAILED",
          error_message: message,
          duration_ms: durationMs,
          completed_at: failedAt,
        }
      );
      this.ledger.appendEvent(
        ctx.executionId,
        EventType.ACTION_FAILED,
        ctx.agentId,
        { error: message },
        failedAt
      );
      throw err;
    }
  }
}

File: rust/src/execution_engine.rs
use crate::approval::{prompt_cli_approval, ApprovalRequest};
use crate::error::{
    SqubeApprovalPendingError, SqubeBlockedError, SqubeDeniedError, SqubeGuardError,
};
use crate::ledger::Ledger;
use crate::models::{ActionRecord, ActionStatus, Decision};
use chrono::{Duration, Utc};
use sha2::{Digest, Sha256};
use uuid::Uuid;

pub type PolicyFn = fn(&str, Option<&str>, &str) -> Decision;
pub type ApprovalCallback = fn(&ApprovalRequest) -> (bool, Option<String>, Option<String>);

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum ApprovalMode {
    Sync,
    Deferred,
}

pub struct ExecutionContext {
    pub execution_id: String,
    pub agent_id: String,
    pub action: String,
    pub resource: Option<String>,
    pub parameters_payload: String,
}

pub struct ExecutionEngine {
    ledger_path: String,
    policy: PolicyFn,
    approval_mode: ApprovalMode,
    approval_timeout_seconds: u64,
    approval_fn: ApprovalCallback,
    policy_id: String,
}

impl ExecutionEngine {
    pub fn new(policy: PolicyFn) -> Self {
        Self {
            ledger_path: "sqube_ledger.sqlite3".to_string(),
            policy,
            approval_mode: ApprovalMode::Sync,
            approval_timeout_seconds: 300,
            approval_fn: prompt_cli_approval,
            policy_id: "policy".to_string(),
        }
    }

    pub fn with_ledger_path(mut self, path: &str) -> Self {
        self.ledger_path = path.to_string();
        self
    }

    pub fn with_approval_mode(mut self, mode: ApprovalMode) -> Self {
        self.approval_mode = mode;
        self
    }

    pub fn with_approval_fn(mut self, f: ApprovalCallback) -> Self {
        self.approval_fn = f;
        self
    }

    pub fn with_policy_id(mut self, id: &str) -> Self {
        self.policy_id = id.to_string();
        self
    }

    pub fn latest_status(&self) -> rusqlite::Result<Option<String>> {
        Ledger::new(&self.ledger_path)?.latest_status()
    }

    pub fn run_controlled<F, R>(
        &self,
        ctx: &ExecutionContext,
        f: F,
    ) -> Result<R, Box<dyn std::error::Error + Send + Sync>>
    where
        F: FnOnce() -> Result<R, Box<dyn std::error::Error + Send + Sync>>,
    {
        let ledger = Ledger::new(&self.ledger_path)?;
        let created_at = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
        let parameters_hash = hex_hash(&ctx.parameters_payload);

        ledger.insert(&ActionRecord {
            execution_id: ctx.execution_id.clone(),
            created_at: created_at.clone(),
            agent_id: ctx.agent_id.clone(),
            action: ctx.action.clone(),
            resource: ctx.resource.clone(),
            parameters_hash,
            parameters_summary: None,
            decision: Decision::Allow.as_str().to_string(),
            policy_id: self.policy_id.clone(),
            status: ActionStatus::Requested.as_str().to_string(),
            approved_by: None,
            approval_reason: None,
            result_status: None,
            error_message: None,
            duration_ms: None,
            completed_at: None,
        })?;

        ledger.transition_execution(
            &ctx.execution_id,
            ActionStatus::Requested,
            ActionStatus::Evaluated,
        )?;

        let decision = (self.policy)(&ctx.action, ctx.resource.as_deref(), &ctx.agent_id);
        ledger.set_decision(&ctx.execution_id, decision.as_str(), ActionStatus::Evaluated.as_str())?;

        match decision {
            Decision::Block => {
                ledger.transition_execution(
                    &ctx.execution_id,
                    ActionStatus::Evaluated,
                    ActionStatus::Blocked,
                )?;
                ledger.set_completed(&ctx.execution_id, &Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true))?;
                return Err(Box::new(SqubeBlockedError {
                    execution_id: ctx.execution_id.clone(),
                    action: ctx.action.clone(),
                }));
            }
            Decision::RequireApproval => {
                if self.approval_mode == ApprovalMode::Deferred {
                    return self.deferred_approval(&ledger, ctx);
                }
                return self.sync_approval(&ledger, ctx, f);
            }
            Decision::Allow => {}
        }

        self.begin_execute(&ledger, ctx, f, ActionStatus::Evaluated)
    }

    pub fn resume_after_approval<F, R>(
        &self,
        ctx: &ExecutionContext,
        approval_id: &str,
        f: F,
    ) -> Result<R, Box<dyn std::error::Error + Send + Sync>>
    where
        F: FnOnce() -> Result<R, Box<dyn std::error::Error + Send + Sync>>,
    {
        let ledger = Ledger::new(&self.ledger_path)?;
        let row = ledger
            .get_execution(&ctx.execution_id)?
            .ok_or_else(|| SqubeGuardError("unknown execution".into()))?;
        if row.status == ActionStatus::WaitingApproval.as_str() {
            return Err(Box::new(SqubeApprovalPendingError {
                execution_id: ctx.execution_id.clone(),
                approval_id: approval_id.to_string(),
            }));
        }
        let hash = hex_hash(&ctx.parameters_payload);
        if hash != row.parameters_hash {
            return Err(Box::new(SqubeGuardError(
                "parameters changed after authorization".into(),
            )));
        }
        let now = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
        ledger.consume_approval(approval_id, &ctx.execution_id, &now)?;
        self.run_fn_body(&ledger, &ctx.execution_id, f)
    }

    fn deferred_approval<R>(
        &self,
        ledger: &Ledger,
        ctx: &ExecutionContext,
    ) -> Result<R, Box<dyn std::error::Error + Send + Sync>> {
        ledger.transition_execution(
            &ctx.execution_id,
            ActionStatus::Evaluated,
            ActionStatus::WaitingApproval,
        )?;
        let approval_id = format!("sq_apr_{}", Uuid::new_v4().simple().to_string()[..12].to_string());
        let now = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
        let expires = (Utc::now() + Duration::seconds(self.approval_timeout_seconds as i64))
            .to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
        ledger.create_approval_request(
            &approval_id,
            &ctx.execution_id,
            &hex_hash(&ctx.parameters_payload),
            &now,
            Some(&expires),
        )?;
        Err(Box::new(SqubeApprovalPendingError {
            execution_id: ctx.execution_id.clone(),
            approval_id,
        }))
    }

    fn sync_approval<F, R>(
        &self,
        ledger: &Ledger,
        ctx: &ExecutionContext,
        f: F,
    ) -> Result<R, Box<dyn std::error::Error + Send + Sync>>
    where
        F: FnOnce() -> Result<R, Box<dyn std::error::Error + Send + Sync>>,
    {
        ledger.transition_execution(
            &ctx.execution_id,
            ActionStatus::Evaluated,
            ActionStatus::WaitingApproval,
        )?;
        let summary = ctx.parameters_payload.clone();
        let (approved, approved_by, reason) = (self.approval_fn)(&ApprovalRequest {
            execution_id: ctx.execution_id.clone(),
            agent_id: ctx.agent_id.clone(),
            action: ctx.action.clone(),
            resource: ctx.resource.clone(),
            summary,
        });
        if !approved {
            let status = if reason.as_deref() == Some("timeout") {
                ActionStatus::Expired
            } else {
                ActionStatus::Denied
            };
            ledger.transition_execution(
                &ctx.execution_id,
                ActionStatus::WaitingApproval,
                status,
            )?;
            ledger.set_denied_meta(
                &ctx.execution_id,
                reason.as_deref().unwrap_or("denied"),
                &Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
            )?;
            return Err(Box::new(SqubeDeniedError {
                execution_id: ctx.execution_id.clone(),
                reason: reason.unwrap_or_else(|| "denied".to_string()),
            }));
        }
        ledger.transition_execution(
            &ctx.execution_id,
            ActionStatus::WaitingApproval,
            ActionStatus::Approved,
        )?;
        ledger.set_approved_by(
            &ctx.execution_id,
            approved_by.as_deref().unwrap_or("cli_user"),
        )?;
        self.begin_execute(ledger, ctx, f, ActionStatus::Approved)
    }

    fn begin_execute<F, R>(
        &self,
        ledger: &Ledger,
        ctx: &ExecutionContext,
        f: F,
        from: ActionStatus,
    ) -> Result<R, Box<dyn std::error::Error + Send + Sync>>
    where
        F: FnOnce() -> Result<R, Box<dyn std::error::Error + Send + Sync>>,
    {
        ledger.transition_execution(&ctx.execution_id, from, ActionStatus::Executing)?;
        self.run_fn_body(ledger, &ctx.execution_id, f)
    }

    fn run_fn_body<F, R>(
        &self,
        ledger: &Ledger,
        execution_id: &str,
        f: F,
    ) -> Result<R, Box<dyn std::error::Error + Send + Sync>>
    where
        F: FnOnce() -> Result<R, Box<dyn std::error::Error + Send + Sync>>,
    {
        let start = std::time::Instant::now();
        match f() {
            Ok(value) => {
                let duration_ms = start.elapsed().as_millis() as i64;
                let completed = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
                ledger.transition_execution(execution_id, ActionStatus::Executing, ActionStatus::Succeeded)?;
                ledger.mark_success_meta(execution_id, duration_ms, &completed)?;
                Ok(value)
            }
            Err(err) => {
                let duration_ms = start.elapsed().as_millis() as i64;
                let completed = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
                ledger.transition_execution(execution_id, ActionStatus::Executing, ActionStatus::Failed)?;
                ledger.mark_failed(
                    execution_id,
                    &err.to_string(),
                    Some(duration_ms),
                    &completed,
                )?;
                Err(err)
            }
        }
    }
}

fn hex_hash(payload: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(payload.as_bytes());
    format!("{:x}", hasher.finalize())
}
